### PR TITLE
SITL: Add C++ library for JSON interface

### PR DESCRIPTION
**Pull request by TunaLobster**
We were working on creating a simulator from scratch to meet some specific needs and I realized there wasn't a plug-and-play way to add support for ArduPilot to a new sim. This is probably trivial for most, but I found this to be very useful for my work.

Right now this code is not the best and naming is open for suggestions. Ideally, there would be a proper way to change what pieces of JSON data to send to ArduPilot, but I am not sure if that should be done with API methods with default value NULL values and several statements similar to `if airseed is not None`. I know in Python I would use optional arguments, but I'm not sure what something similar would look like here.

### CODE REVIEW

**Comment by TunaLobster** [comment]
Tested working on WSL 1 and 2 Ubuntu 20.04.

**Comment by davidbuzz** [comment]
just curious, its still marked as wIP, but i think it should be mergable, yes..?  @TunaLobster ?

**Comment by TunaLobster** [comment]
It works and gets the job done. It's not as complete as I would want it to be though. Testing with winsock and a better way to set the optional JSON interface values would be nice.

**Comment by TunaLobster** [comment]
Added setters for windvane and airspeed. Screenshot is the whole thing running through WSL 1. Cleaned up some pieces too.

@davidbuzz I'll call it good for right now. Getting winsock and rangefinder is getting pushed down the road on my schedule.

File: libraries/SITL/examples/JSON/C++/Socket.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  simple socket handling class for systems with BSD socket API
+ */
```
Comment (peterbarker) [inline libraries/SITL/examples/JSON/C++/Socket.h:17]: Remove

File: libraries/SITL/examples/JSON/C++/Socket.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#pragma once
+
+// #include <AP_HAL/AP_HAL.h>
+// #if HAL_OS_SOCKETS
```
Comment (peterbarker) [inline libraries/SITL/examples/JSON/C++/Socket.h:21]: Remove

File: libraries/SITL/examples/JSON/C++/Socket.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <arpa/inet.h>
+#include <sys/select.h>
+
+class SocketAPM {
```
Comment (peterbarker) [inline libraries/SITL/examples/JSON/C++/Socket.h:34]: Rename

Move to C++ code as being an internal thing?

Comment (tridge) [inline libraries/SITL/examples/JSON/C++/Socket.h:34]: can we please rename this to a different name than the AP_HAL/utility/Socket.h

**Review by tridge — CHANGES_REQUESTED** [review]
add a README.md on how to build and test this

File: libraries/SITL/examples/JSON/C++/libAP_JSON.cpp [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    // then socket receive wait time is increased from 1ms to 1 sec
+    // to accommodate network jitter.
+    // If ArduPilot is not detected, receive call blocks for 1ms
+    // on each call.
+    // Once ArduPilot presence is detected, it takes this many
+    // missed receives before declaring the FCS offline.
+
+    uint32_t waitMs;
+    if (arduPilotOnline)
+    {
```
Comment (peterbarker) [inline libraries/SITL/examples/JSON/C++/libAP_JSON.cpp:71]:
```suggestion
    if (arduPilotOnline) {
```


... and generally ardupilot-ise

**Review by peterbarker — CHANGES_REQUESTED** [review]
Short in-meeting review

File: libraries/SITL/examples/JSON/C++/Socket.h (hunk 3 above)
Comment (TunaLobster) [reply libraries/SITL/examples/JSON/C++/Socket.h:34]: Fixed.

File: libraries/SITL/examples/JSON/C++/Socket.h (hunk 1 above)
Comment (TunaLobster) [reply libraries/SITL/examples/JSON/C++/Socket.h:17]: Fixed.

File: libraries/SITL/examples/JSON/C++/Socket.h (hunk 2 above)
Comment (TunaLobster) [reply libraries/SITL/examples/JSON/C++/Socket.h:21]: Fixed.

File: libraries/SITL/examples/JSON/C++/Socket.h (hunk 3 above)
Comment (TunaLobster) [reply libraries/SITL/examples/JSON/C++/Socket.h]: Renamed to SocketExample

File: libraries/SITL/examples/JSON/C++/libAP_JSON.cpp (hunk 4 above)
Comment (TunaLobster) [reply libraries/SITL/examples/JSON/C++/libAP_JSON.cpp]: Fixed up the ones I recognize.

**Comment by TunaLobster** [comment]
A readme and a 1-D have been added. Fixed an issue where the rangefinder data is properly indexed at 1 rather than 0.

**Review by tridge — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**